stan-dev/cmdstan
# Fixing the makefiles

#### Submisison Checklist


#### Summary:
This pull request simplifies the build and gets the makefiles in order. The makefiles got tangled up through a few recent changes. This uses changes from the PR in Math: stan-dev/math#956, and the PR in Stan: stan-dev/stan#2601.

Here are some of the highlights:
- `CXX` is the only compiler setting used. In develop, both `CXX` and `CC` have to be set
- Windows: removed the `uname: not found` message
- `CXXFLAGS` are now broken into smaller components that can be set by the user:
    - `CXXFLAGS_LANG` for language-specific flags
    - `CXXFLAGS_WARNINGS` for ignoring warnings
    - `CXXFLAGS_BOOST` for boost-specific flags
    - `CXXFLAGS_SUNDIALS` for sundials-specific flags
    - `CXXFLAGS_EIGEN` for eigen-specific flags
    - `CXXFLAGS_GTEST` for google-test-specific flags
    - `CXXFLAGS_OS` for a catch-all and any os-specific flags.
- `CPPFLAGS` are broken into components (`CPPFLAGS` are the C preprocessor flags):
    - `CPPFLAGS_BOOST` for boost-specific flags
    - `CPPFLAGS_SUNDIALS` for sundials-specific flags
    - `CPPFLAGS_EIGEN` for eigen-specific flags
    - `CPPFLAGS_GTEST` for google-test-specific flags
    - `CPPFLAGS_OPENCL` for opencl

  breaking apart the flags this way makes it clearer how to extend the flags for MPI and GPU.
- each of the flags can be set within `make/local` and it will take precedence over the default values
- the dependency generation now generates with dummy targets for each file. This means more robustness in that -- this fixes problems where a header file is removed from the source tree. This caused a lot of trouble with stability of make.

#### Intended Effect:

This will get it so the makefiles shouldn't really need updating for a while. It's modular enough for extension with MPI or GPU or other will be simple.

#### How to Verify:

1. Run all tests. Everything should build. @bbbales2, I even got it so we can update one of the files in the parser and it'll rebuild the appropriate tests.
2. Verify that you can build a new Stan executable using these makefiles.

#### Side Effects:

We no longer use `CC`. This will have to be updated to `CXX`.

#### Documentation:

- [x] After this is working, update the wiki and the manual appropriately.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Generable

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)
